In [1]:
import numpy as np
import pandas as pd

第一題

In [2]:
df = pd.read_csv("economic_freedom_data.csv")
df

government_consumption  transfers  gov_enterprises  top_marg_tax_rate  \
0                   8.232353   7.509902         8.000000                8.0   
1                   2.150000   7.817129         0.000000                4.5   
2                   7.600000   8.886739         0.000000                9.5   
3                   5.335294   6.048930         6.000000                4.0   
4                   7.264706   7.748532         8.000000                5.0   
...                      ...        ...              ...                ...   
1795                7.884038   6.600000         7.000000                7.5   
1796                6.361765   5.773842         2.000000                5.5   
1797                9.166904   7.650169         6.310795                4.0   
1798                4.823529  10.000000         4.000000                8.0   
1799                7.600000   7.746594         0.000000                3.0   

      size_government  judicial_independence  impartial_courts  \
0            7.935564               2.668222          3.145462   
1            3.616782               4.186704          4.327113   
2            6.496685               1.843129          1.974566   
3            5.346056               3.689690          2.930563   
4            7.003310               3.867379          4.197569   
...               ...                    ...               ...   
1795         7.246010               6.550926          5.469484   
1796         4.908902               0.307692          0.934343   
1797         6.583452               4.133858          4.603175   
1798         6.705882               2.859649          5.000000   
1799         4.586649               1.095238          2.142857   

      protection_property_rights  military_interference  \
0                       4.512228               8.333333   
1                       4.689952               4.166667   
2                       2.512364               3.333333   
3                       4.255995               7.500000   
4                       5.664317               5.833333   
...                          ...                    ...   
1795                    6.261905               8.194444   
1796                    2.272727               0.833333   
1797                    5.721393               5.000000   
1798                    6.048110               8.333333   
1799                    1.435185               5.000000   

      integrity_legal_system  ...   tariffs  regulatory_trade_barriers  \
0                   4.166667  ...  8.963556                   7.489905   
1                   5.000000  ...  6.872533                   2.481294   
2                   4.166667  ...  6.989244                   2.024949   
3                   3.333333  ...  6.421600                   4.811105   
4                   5.000000  ...  8.547556                   7.194410   
...                      ...  ...       ...                        ...   
1795                4.166667  ...  7.729511                   6.221928   
1796                5.000000  ...  7.640889                   4.039648   
1797                6.666667  ...  5.795427                   5.532874   
1798                6.666667  ...  6.965600                   3.396974   
1799                5.000000  ...  3.924444                   3.148471   

      black_market  control_movement_capital_ppl     trade  credit_market_reg  \
0        10.000000                      6.406138  8.214900           7.098562   
1         5.563910                      1.590362  4.127025           5.100509   
2        10.000000                      2.044823  5.264754           7.064905   
3         0.000000                      4.697482  3.982547           5.419820   
4        10.000000                      6.830998  8.143241           9.102046   
...            ...                           ...       ...                ...   
1795     10.000000                      7.398514  7.837488           6.711135   
1796      5.128086

In [3]:
df.info() #可以看到全部都沒有遺漏值

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1800 entries, 0 to 1799
Data columns (total 31 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   government_consumption           1800 non-null   float64
 1   transfers                        1800 non-null   float64
 2   gov_enterprises                  1800 non-null   float64
 3   top_marg_tax_rate                1800 non-null   float64
 4   size_government                  1800 non-null   float64
 5   judicial_independence            1800 non-null   float64
 6   impartial_courts                 1800 non-null   float64
 7   protection_property_rights       1800 non-null   float64
 8   military_interference            1800 non-null   float64
 9   integrity_legal_system           1800 non-null   float64
 10  legal_enforcement_contracts      1800 non-null   float64
 11  restrictions_sale_real_property  1800 non-null   float64
 12  reliability_police  

In [4]:
df.describe().T #每一個欄位的平均值和標準差都有一定差距，表明都要列為關係欄位

count      mean       std       min  \
government_consumption           1800.0  5.736468  2.263399  0.000000   
transfers                        1800.0  7.650794  2.031597  1.940251   
gov_enterprises                  1800.0  6.310968  3.050884  0.000000   
top_marg_tax_rate                1800.0  6.374873  2.203242  0.000000   
size_government                  1800.0  6.487953  1.286690  2.870588   
judicial_independence            1800.0  4.917326  2.101457  0.000000   
impartial_courts                 1800.0  4.470405  1.616641  0.000000   
protection_property_rights       1800.0  5.684567  1.686075  0.000000   
military_interference            1800.0  6.452639  2.726764  0.000000   
integrity_legal_system           1800.0  6.230758  1.995502  0.166667   
legal_enforcement_contracts      1800.0  4.449870  1.682053  0.000000   
restrictions_sale_real_property  1800.0  7.222416  1.874184  0.000000   
reliability_police               1800.0  5.507285  1.897749  0.000000   
business_costs_crime             1800.0  5.903404  1.785188  0.000000   
gender_adjustment                1800.0  0.891418  0.141039  0.285714   
property_rights                  1800.0  5.303717  1.631387  1.369178   
money_growth                     1800.0  8.466558  1.251559  0.000000   
std_inflation                    1800.0  8.358846  1.775566  0.000000   
inflation                        1800.0  8.856679  1.310712  0.000000   
freedom_own_foreign_currency     1800.0  1.322222  0.823876  0.000000   
sound_money                      1800.0  8.071761  1.423497  0.000000   
tariffs                          1800.0  7.394233  1.251203  3.116622   
regulatory_trade_barriers        1800.0  6.160852  1.707630  0.000000   
black_market                     1800.0  9.758193  1.305530  0.000000   
control_movement_capital_ppl     1800.0  4.875888  2.069321  0.000000   
trade                            1800.0  7.039380  1.144365  1.781689   
credit_market_reg                1800.0  8.354469  1.421763  2.206734   
labor_market_reg                 1800.0  6.388576  1.437468  2.099009   
business_reg                     1800.0  6.229386  1.236700  2.009841   
regulation                       1800.0  6.985895  1.037915  2.483540   

                                       25%        50%        75%        max  
government_consumption            4.367647   5.927941   7.411765  10.000000  
transfers                         6.318801   8.109183   9.379952  10.000000  
gov_enterprises                   4.000000   7.000000   8.000000  10.000000  
top_marg_tax_rate                 5.000000   6.500000   8.000000  10.000000  
size_government                   5.593928   6.537883   7.432712   9.528485  
judicial_independence             3.329316   4.772710   6.458105   9.697464  
impartial_courts                  3.247160   4.156201   5.446937   9.253731  
protection_property_rights        4.549596   5.619270   6.868694   9.607843  
military_interference             4.167000   6.666667   8.333333  10.000000  
integrity_legal_system            5.000000   6.231000   8.333333  10.000000  
legal_enforcement_contracts       3.389827   4.449797   5.583266   8.478714  
restrictions_sale_real_property   6.271657   7.581509   8.574071   9.981101  
reliability_police                4.198208   5.506447   6.789255   9.688242  
business_costs_crime              4.750759   5.903589   7.205075   9.672608  
gender_adjustment                 0.827586   0.948718   1.000000   1.000000  
property_rights                   4.122490   5.241925   6.273862   9.138119  
money_growth                      8.046632   8.715705   9.251070   9.999031  
std_inflation                     7.980891   9.001117   9.460847   9.952306  
inflation                         8.500186   9.223704   9.635496  10.000000  
freedom_own_foreign_currency      1.000000   2.000000   2.000000   2.000000  
sound_money                       7.005633   8.290855   9.351003   9.922187  
tariffs                           6.598711   7.641556 

In [5]:
df['freedom'].unique() #可以看到target是object 並只有兩個值，freedom 和 not-freedom

array(['freedom', 'not-freedom'], dtype=object)

In [6]:
df['freedom'] = df['freedom'].replace('not-freedom', 0).replace('freedom', 1) #替換freedom為1 not-freedom為0

In [7]:
df_shuffled = df.sample(frac=1).reset_index(drop=True) #將排序洗牌

In [8]:
y = df_shuffled['freedom']
X = df_shuffled.drop(['freedom'], axis=1) #分為target和關係欄位

第二題

In [9]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import f1_score

In [10]:
# RandomForest
rfc = RandomForestClassifier(random_state=10)
rfc_f1_scores = cross_val_score(rfc, X, y, cv=10)
average_rfc_f1_score = rfc_f1_scores.mean()
print("RandomForest Average F1 Score:", average_rfc_f1_score)

RandomForest Average F1 Score: 0.9683333333333334


第三題

In [11]:
from sklearn.neural_network import MLPClassifier
mlp_classifier = MLPClassifier(hidden_layer_sizes=(64, 128), random_state=10, batch_size=8)
mlp_f1_scores = cross_val_score(mlp_classifier, X, y, cv=10, scoring='f1_macro')
average_mlp_f1_score = mlp_f1_scores.mean()
print("Scikit-learn MLP Average F1 Score:", average_mlp_f1_score)

Scikit-learn MLP Average F1 Score: 0.9519851752333377


第四題

In [12]:
import tensorflow
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import StratifiedKFold


In [13]:
def create_model():
    model = Sequential()
    model.add(Dense(64, input_dim=X.shape[1], activation='relu'))
    model.add(Dense(128, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model


kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=10)
k_mlp_f1_scores = []
for train_index, test_index in kfold.split(X, y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]


    model = create_model()


    model.fit(X_train, y_train, epochs=10, batch_size=8, verbose=0)

    y_pred_prob = model.predict(X_test)
    y_pred = np.where(y_pred_prob > 0.5, 1, 0)

    f1 = f1_score(y_test, y_pred)


    k_mlp_f1_scores.append(f1)


# 打印平均 F1 分数
average_k_mlp_f1_score = np.mean(k_mlp_f1_scores)
print("Keras MLP Average F1 Score:", average_k_mlp_f1_score)

6/6 [==============================] - 0s 3ms/step
Keras MLP Average F1 Score: 0.9163851566676497


第五題

freedom ,f1score是如果沒有指定是會將freedom也就是正向的值當作依據

第六題

In [14]:
from scipy.stats import ttest_ind



t_statistic, p_value = ttest_ind(rfc_f1_scores,mlp_f1_scores )
t_statistic2, p_value2 = ttest_ind(rfc_f1_scores, k_mlp_f1_scores)
t_statistic3, p_value3 = ttest_ind(mlp_f1_scores,k_mlp_f1_scores )
 
print("Comparison between RandomForest and scikit-learn MLP:")
print("t-statistic:", t_statistic)
print("p-value:", p_value)
print()
print("Comparison between RandomForest and Keras MLP:")
print("t-statistic:", t_statistic2)
print("p-value:", p_value2)
print()
print("Comparison between scikit-learn MLP and Keras MLP:")
print("t-statistic:", t_statistic3)
print("p-value:", p_value3)

Comparison between RandomForest and scikit-learn MLP:
t-statistic: 1.993835331813589
p-value: 0.06154624709235281

Comparison between RandomForest and Keras MLP:
t-statistic: 4.617896154977763
p-value: 0.00021367632798739768

Comparison between scikit-learn MLP and Keras MLP:
t-statistic: 2.977507680317037
p-value: 0.008070377452089517


以95%的信心水準來看
隨機森林（RandomForest）和 scikit-learn 的多層感知器（MLP）之間的比較：
t 統計值為 1.993835331813589，p 值為 0.06154624709235281。p>0.05
根據統計結果，我們無法拒絕兩者在表現上沒有顯著差異的假設。

隨機森林（RandomForest）和 Keras 的多層感知器（MLP）之間的比較：
t 統計值為 4.617896154977763，p 值為 0.00021367632798739768。
根據統計結果，我們可以拒絕兩者在表現上沒有顯著差異的假設。這意味著隨機森林和 Keras 的 MLP 在某些方面上可能有顯著的差異。

scikit-learn 的多層感知器（MLP）和 Keras 的多層感知器（MLP）之間的比較：
t 統計值為 2.977507680317037，p 值為 0.008070377452089517。
根據統計結果，我們可以拒絕兩者在表現上沒有顯著差異的假設。這意味著 scikit-learn 的 MLP 和 Keras 的 MLP 在某些方面上可能有顯著的差異。

綜合以上結果，根據所得的統計結果，我們可以得出結論：隨機森林與 scikit-learn 的 MLP 在表現上可能沒有顯著差異；而隨機森林和 Keras 的 MLP 以及 scikit-learn 的 MLP 和 Keras 的 MLP 在表現上可能存在顯著差異。

再看到個別準確率，可以發現隨機森林最高，應為表現最好的模型

第七題

In [15]:
data = pd.read_csv("newdata.csv")#導入測試資料
data

government_consumption  transfers  gov_enterprises  top_marg_tax_rate  \
0                6.473529   6.524018                4                  7   

   size_government  judicial_independence  impartial_courts  \
0         6.499387               1.832149          2.353257   

   protection_property_rights  military_interference  integrity_legal_system  \
0                    3.839514               6.666667                6.666667   

   ...  sound_money   tariffs  regulatory_trade_barriers  black_market  \
0  ...     7.983343  8.681211                   7.097618             9   

   control_movement_capital_ppl     trade  credit_market_reg  \
0                      4.048331  7.458865           9.517982   

   labor_market_reg  business_reg  regulation  
0          5.217961      5.980311    6.905421  

[1 rows x 30 columns]

In [16]:
import joblib
mlp_classifier.fit(X, y)
joblib.dump(mlp_classifier, 'mlp_model.joblib')
loaded_model = joblib.load('mlp_model.joblib')
predictions = loaded_model.predict(data)
predictions #not freedom

array([0], dtype=int64)

In [17]:
rfc.fit(X, y)
joblib.dump(rfc, 'rfc_model.joblib')
loaded_model2 = joblib.load('rfc_model.joblib')
predictions2 = loaded_model2.predict(data)
predictions2

array([0], dtype=int64)

In [18]:
y_pred_prob_new = model.predict(data)
y_pred_new = np.where(y_pred_prob_new > 0.5, 1, 0)

print("Predicted Labels for X_new:", y_pred_new)


1/1 [==============================] - 0s 16ms/step
Predicted Labels for X_new: [[0]]


三個模型都是 not-freedom

第八題

In [19]:
rfc.fit(X,y)
importances = rfc.feature_importances_

In [35]:
feature_importance = dict(zip(X.columns,importances))
np.argsort(feature_importance)[::-1]
sorted_importance = sorted(feature_importance.items(), key=lambda x: x[1], reverse=True)
i=1
for feature, importance in sorted_importance:
    print("(%2d) %-*s %f" %(i,30,feature,importance))
    i+=1

( 1) sound_money                    0.155908
( 2) trade                          0.155640
( 3) freedom_own_foreign_currency   0.101123
( 4) control_movement_capital_ppl   0.072366
( 5) regulation                     0.057650
( 6) business_reg                   0.053318
( 7) property_rights                0.051707
( 8) tariffs                        0.045642
( 9) regulatory_trade_barriers      0.038904
(10) size_government                0.035960
(11) military_interference          0.029500
(12) protection_property_rights     0.022835
(13) impartial_courts               0.015224
(14) labor_market_reg               0.014650
(15) restrictions_sale_real_property 0.013564
(16) gov_enterprises                0.013134
(17) std_inflation                  0.012707
(18) government_consumption         0.012122
(19) credit_market_reg              0.012003
(20) judicial_independence          0.011827
(21) reliability_police             0.010324
(22) business_costs_crime           0.009894
(23) lega

In [ ]:
最重要的是sound_money